### ISSUE TITLE
Regression in Gutenberg 11 around pasting with spaces?

### ISSUE THREAD

**Issue by pansamolot** [WordPress/gutenberg#33283]
## Description
Gutenberg 11.0.0 was auto-updated this morning to my blog network. And I again see problem reported a 1+ year ago about adding surrounding spaces when pasting a text.

I have a feeling that this is at least 3rd time _deja vu_ in Gutenberg. Are there any reported regression that is causing again something that was year ago reported as resolved?

## Step-by-step reproduction instructions

1. Start a new blog post.
2. Add a new Paragraph block.
3. Start typing.
4. Paste some text from the same article or from outer source.

## Expected behaviour
Pasted text is exactly the same as copied from source.

## Actual behaviour
Pasted text is surrounded with spaces.

## Screenshots

![spaces](https://user-images.githubusercontent.com/60975060/124989844-df5e6900-e03f-11eb-98a2-7a41d784bf54.gif)

## WordPress information
- WordPress version: **5.7.2**
- Gutenberg version: **11.0.0**
- Are all plugins except Gutenberg deactivated? **No**
- Are you using a default theme (e.g. Twenty Twenty-One)? **No**

## Device information
- Device: **Notebook PC, i7-7600U CPU, 8 GB RAM**, brand new, very few apps installed, non except browser working,
- Operating system: **Windows 10.10094**
- Browser: **Microsoft Edge 91.0.864.64**

**Comment by getdave** [comment]
hi @pansamolot I just attempt to replicate this and was unable to succeed? What did I miss?

https://user-images.githubusercontent.com/444434/126506855-ee80556c-521d-4c23-8043-69829c255f23.mp4

Also can I verify you aren't using any WordPress Plugins, browser extensions or system level modifications that might alter your OS's pasting mechanics?

If it still happens for you, then perhaps you could try this on a clean install using a clean browser to see if that solves things? You've probably already done this but I felt it was worth checking as I've seen this be the case for other similarly reported issues.

If there is a repeated problem with this and we can replicate then we will look to cover with unit tests.

Much appreciated.

**Comment by pansamolot** [comment]
I have just quickly setup _as-clean-as-possible_ installation of Wordpress on Amazon Lightsail (to throw out any possible connections to my main hosting's configuration).

It has **nothing** except:

* Wordpress 5.8
* Twenty Twenty-One theme 1.3
* Gutenberg plugin 11.1.0
* single user
* single blog post

And I **can confirm** that this bug exits even from the very first line of text written. At this moment I am on another device, so I cannot create an animated gif to show this, but I hope to post it in a couple of hours.

I am not aware of any system settings, browser extensions etc. that can influence my "pasting experience". Wordpress is the only web service that suffers this problem for me even though I am using the same PC, system and browser for pasting texts in dozens of services. The only other exception is Gmail, but it is a well known, years-long bug in Gmail itself that it adds spaces around pasted text.

**Comment by getdave** [comment]
Thanks for confirming - much appreciated. Let's see if we can get other folks to verify the bug and also perhaps some Windows users.

In the meantime any more detail and examples you can provide including those animated GIFs will really help.

**Comment by getdave** [comment]
I finally managed to get a BrowserStack login and gave this a spin on Windows 10 with Edge 91 and was able to replicate:

https://user-images.githubusercontent.com/444434/126521867-fdb613d2-6895-4490-bedd-7476fe09c83b.mov

**Comment by getdave** [comment]
I can confirm it is the call to `create()` below that causes the spaces to be added. 

https://github.com/WordPress/gutenberg/blob/059339d140f7be82ce52fa4eac559678f75e5ca5/packages/block-editor/src/components/rich-text/use-paste-handler.js#L110-L116

The copied content is correct upon copy but seems to be surrounded by HTML "fragment" comments. I wonder if these are somehow causing the issue?

<img width="287" alt="Screenshot 2021-07-21 at 17 10 55" src="https://user-images.githubusercontent.com/444434/126522647-7b781592-c9c7-4a6a-9fdb-cf1d19e5c976.png">

**Comment by t-hamano** [comment]
I tested in the following environment:

# Development Environment

Install npm package only `@wordpress/env`
`.wp-env` is set by default as follows

```
{
  "core": null,
  "themes": [ "." ],
  "config": {
    "WP_DEBUG": true
  }
}
```

# WordPress information
WordPress version: `5.8`
Gutenberg version: `10.7`（without Gutenberg Plugin）
Are you using a default theme (e.g. Twenty Twenty-One)? Yes

# Device information
Device: Notebook PC, i5-1035G1 CPU, 16 GB RAM
Operating system: Windows 10 Home 64bit
Browser: Google Chrome 91.0.4472.164 and Microsoft Edge 91.0.864.71

# Test Results

1. Copy and paste text from **the windows notepad** into **the windows notepad**: Of course no space will be added.
2. Copy and paste text from **the windows notepad** into **the block editor**: no space will be added.
3. Copy and paste text from **the block editor** into **the windows notepad**: no space will be added.
4. Copy and paste text from **the block editor** into **the block editor**: A space will be inserted at both ends of the text.


https://user-images.githubusercontent.com/54422211/126529096-b90639be-3fb6-41a8-a2d7-a19adeddc6a3.mp4

A space will be inserted at both ends of the text only when **I copy the text in the block editor and paste it in the same block editor**.
The results were also the same for both Google Chrome and Windows Edge.

**Comment by t-hamano** [comment]
I don't know the cause as I don't have a deep understanding of Gutenberg.
However, I think that the following article may be the same case as this symptom.
I hope you find it helpful.
https://discuss.prosemirror.net/t/space-added-on-paste/1274

**Comment by pansamolot** [comment]
Thank you @getdave and @t-hamano for all the tests. It seems that I don't need to add a proof from my side as this is already confirmed by you many times.

Seems that the only thing we're left with is to wait for the dev team to pick this up. Thanks for your effort.

**Comment by getdave** [comment]
You are welcome.

I think @t-hamano has it with this post which has details on that is going on with this fragment business.

https://discuss.prosemirror.net/t/space-added-on-paste/1274/7

As a member of the dev team, I'll see if I can look into a fix tomorrow.

**Update**: ok I lied. I have a fix PR available now in https://github.com/WordPress/gutenberg/pull/33607. It's not ready for full review yet but it should work to fix the issue.

**Comment by pansamolot** [comment]
Blasting, light speed and amazingly fast! :> Thank you! :>

**Comment by getdave** [comment]
@pansamolot Would you be able to try and [test the PR](https://github.com/WordPress/gutenberg/pull/33607) for me? You can use [these instructions](https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/) to do this pretty easily and quickly.

**Comment by pansamolot** [comment]
@getdave Just made a quick test of your PR on a brand new AWS Wordpress instance and from my perspective the issue is solved. 

There are no spaces added in any scenario where they're not expected and there are spaces added when they are expected (i.e. when copy-pasting a piece of text that actually contains beginning and ending space).

Due to the limited time resources I only managed to test this by making various pieces of text around the same blog post in Wordpress. I did not had a chance to test external-to-Wordpress or Wordpress-to-external copy-paste, but that was working even previously, so I assumed that your PR did not changed anything around here.

Thank you, great job!
